tests/src/Kernel/Engine/MetadataAssertionTest.php

```diff
@@ -0,0 +1,95 @@
+<?php
+
+/**
+ * @file
+ * Contains \Drupal\Tests\rules\Kernel\Engine\MetadataAssertionTest.
+ */
+
+namespace Drupal\Tests\rules\Kernel\Engine;
+
+use Drupal\field\Entity\FieldConfig;
+use Drupal\field\Entity\FieldStorageConfig;
+use Drupal\rules\Context\ContextConfig;
+use Drupal\rules\Context\ContextDefinition;
+use Drupal\rules\Engine\RulesComponent;
+use Drupal\Tests\rules\Kernel\RulesDrupalTestBase;
+
+/**
+ * Tests asserting metadata works correctly.
+ *
+ * @group rules
+ */
+class MetadataAssertionTest extends RulesDrupalTestBase {
+
+  /**
+   * {@inheritdoc}
+   */
+  public static $modules = ['rules', 'system', 'node', 'field', 'user'];
+
+  /**
+   * {@inheritdoc}
+   */
+  public function setUp() {
+    parent::setUp();
+
+    $this->installSchema('system', ['sequences']);
+    $this->installEntitySchema('user');
+    $this->installEntitySchema('node');
+    $this->installConfig(['field']);
+
+    $entity_type_manager = $this->container->get('entity_type.manager');
+    $entity_type_manager->getStorage('node_type')
+      ->create(['type' => 'page'])
+      ->save();
+
+    FieldStorageConfig::create([
+      'field_name' => 'field_text',
+      'type' => 'string',
+      'entity_type' => 'node',
+      'cardinality' => 1,
+    ])->save();
+    FieldConfig::create([
+      'field_name' => 'field_text',
+      'entity_type' => 'node',
+      'bundle' => 'page',
+    ])->save();
+  }
+
+  /**
+   * Tests asserting metadata using the EntityIfOfBundle condition.
+   */
+  public function testAssertingEntityBundle() {
+    // When trying to use the field_text field without knowledge of the bundle,
+    // the field is not available.
+    $rule = $this->expressionManager->createRule();
+    $rule->addAction('rules_system_message', ContextConfig::create()
+      ->map('message', 'node.field_text.value')
+      ->setValue('type', 'status')
+    );
+    $violation_list = RulesComponent::create($rule)
+      ->addContextDefinition('node', ContextDefinition::create('entity:node'))
+      ->checkIntegrity();
+    $this->assertEquals(1, iterator_count($violation_list));
+    $this->assertEquals(
+      'Data selector %selector for context %context_name is invalid. @message',
+      $violation_list->get(0)->getMessage()->getUntranslatedString()
+    );
+
+    // Now add the EntityIsOfBundle condition and try again.
+    $rule = $this->expressionManager->createRule();
```

Comment:
Why do you have to create a new rule? Just adding the condition to the existing one should be enough?